stl/inc/functional

```diff
@@ -1207,6 +1207,604 @@ _NODISCARD bool operator!=(nullptr_t, const function<_Fty>& _Other) noexcept {
+public:
+    using result_type = _Rx;
+
+    template <class _Vt>
+    using _VtInvQuals = _Vt&&;
+
+    template <class _Vt>
+    static constexpr bool _Is_callable_from = is_nothrow_invocable_r_v<_Rx, _Vt&&, _Types...>;
+
+    _Rx operator()(_Types... _Args) && noexcept {
+        return this->_Get_invoke()(this->_Data, _STD forward<_Types>(_Args)...);
+    }
+};
+
+template <class _Rx, class... _Types>
+class _Move_only_function_call<_Rx(_Types...) const noexcept> : public _Move_only_function_base<_Rx, _Types...> {
+public:
+    using result_type = _Rx;
+
+    template <class _Vt>
+    using _VtInvQuals = _Vt const&;
+
+    template <class _Vt>
+    static constexpr bool _Is_callable_from =
+        is_nothrow_invocable_r_v<_Rx, _Vt const, _Types...>&& is_nothrow_invocable_r_v<_Rx, _Vt const&, _Types...>;
+
+    _Rx operator()(_Types... _Args) const noexcept {
+        return this->_Get_invoke()(this->_Data, _STD forward<_Types>(_Args)...);
+    }
+};
+
+template <class _Rx, class... _Types>
+class _Move_only_function_call<_Rx(_Types...) const& noexcept> : public _Move_only_function_base<_Rx, _Types...> {
+public:
+    using result_type = _Rx;
+
+    template <class _Vt>
+    using _VtInvQuals = _Vt const&;
+
+    template <class _Vt>
+    static constexpr bool _Is_callable_from = is_nothrow_invocable_r_v<_Rx, _Vt const&, _Types...>;
+
+    _Rx operator()(_Types... _Args) const& noexcept {
+        return this->_Get_invoke()(this->_Data, _STD forward<_Types>(_Args)...);
+    }
+};
+
+template <class _Rx, class... _Types>
+class _Move_only_function_call<_Rx(_Types...) const&& noexcept> : public _Move_only_function_base<_Rx, _Types...> {
+public:
+    using result_type = _Rx;
+
+    template <class _Vt>
+    using _VtInvQuals = _Vt const&&;
+
+    template <class _Vt>
+    static constexpr bool _Is_callable_from = is_nothrow_invocable_r_v<_Rx, _Vt const&&, _Types...>;
+
+    _Rx operator()(_Types... _Args) const&& noexcept {
+        return this->_Get_invoke()(this->_Data, _STD forward<_Types>(_Args)...);
+    }
+};
+#endif // __cpp_noexcept_function_type
+
+template <class _Fx>
+class move_only_function : private _Move_only_function_call<_Fx> {
+private:
+    using _Call = _Move_only_function_call<_Fx>;
+
+    // clang-format off
+    template <class _Fn>
+    static constexpr bool _Enable_one_arg_constructor =
+        !is_same_v<_Remove_cvref_t<_Fn>, move_only_function> &&
+        !_Is_specialization_v<_Remove_cvref_t<_Fn>, in_place_type_t> &&
+        _Call::template _Is_callable_from<decay_t<_Fn>>;
+
+    template <class _Fn, class... _CTypes>
+    static constexpr bool _Enable_in_place_constructor =
+        is_constructible_v<decay_t<_Fn>, _CTypes...> &&
+        _Call::template _Is_callable_from<decay_t<_Fn>>;
+
+    template <class _Fn, class _Ux, class... _CTypes>
+    static constexpr bool _Enable_in_place_list_constructor =
+        is_constructible_v<decay_t<_Fn>, initializer_list<_Ux>, _CTypes...> &&

```

Comment:
This should be `initializer_list<_Ux>&` according to N4901 [func.wrap.move.ctor]/17.1 "`is_constructible_v<VT, initializer_list<U>&, ArgTypes...>` is `true`, and".

Reply:
Why? Initializer list is passed by value.

Reply:
The user's type might care about the difference:

```cpp
#include <initializer_list>
#include <type_traits>
using namespace std;

struct Meow {
    Meow(initializer_list<int>&&, double) = delete;
    Meow(const initializer_list<int>&, double) {}
};

int main() {
    static_assert(!is_constructible_v<Meow, initializer_list<int>, double>);
    static_assert(is_constructible_v<Meow, initializer_list<int>&, double>);
}
```

Reply:
Well, I changed the condition, still not sure why would user's type take `initializer_list` in any other way than by value.

Reply:
It's unlikely, but users are permitted to do a lot of wacky things.